src/box/tx_memory.h

```diff
@@ -0,0 +1,212 @@
+#pragma once
+/*
+ * Copyright 2010-2017, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+ * This file contains declaration class @a tx_stat that is common memory manager
+ * for every transaction manager and txns itself.
+ * Txns and transaction managers must allocate memory using methods @a tx_stat
+ * only because it make possible to monitor memory usage,
+ * and that is why every transaction manager must have its own memory manager
+ * derived from this one.
+ * It is also important to say that one should not create an instance of
+ * @a tx_stat, create instance of derivated memory manager instead.
+ */
+
+#include <trivia/util.h>
+#include "small/mempool.h"
+#include "histogram.h"
+struct txn;
+struct tuple;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Allocations types for txns.
+ */
+enum TXN_ALLOC_TYPE {
+	TXN_ALLOC_STMT = 0,
+	TXN_ALLOC_SVP = 1,
+	TXN_ALLOC_USER_DATA = 2,
+	TXN_ALLOC_REDO_LOG = 3,
+	TXN_ALLOC_TRIGGER = 4,
+	TXN_ALLOC_TIMER = 5,
```

Comment:
Timers, triggers and savepoints are quite rare in terms of allocations. Why not just classify them all as sort of "system"/"metadata"?

Reply:
Agree